include/RAJA/policy/cuda/reduce.hpp

```diff
@@ -109,1507 +190,1036 @@ __device__ __forceinline__ T shfl_xor(T var, int laneMask)
+      : parent{this},
+        info{0},
+        val(init_val)
   {
-    m_is_copy_host = false;
-    m_myID = getCudaReductionId();
-    getCudaReductionMemBlock(m_myID, (void **)&m_blockdata);
-    getCudaReductionTallyBlock(m_myID,
-                               (void **)&m_tally_host,
-                               (void **)&m_tally_device);
-    m_tally_host->tally = init_val;
-    m_tally_host->retiredBlocks = static_cast<GridSizeType>(0);
   }
 
-  /*!
-   * \brief Initialize shared memory on device, request shared memory on host.
-   *
-   * Copy constructor executes on both host and device.
-   * On host requests dynamic shared memory and gets offset into dynamic
-   * shared memory if in forall.
-   * On device initializes dynamic shared memory to appropriate value.
-   */
-  __host__ __device__
-  ReduceSum(const ReduceSum<cuda_reduce<BLOCK_SIZE, Async>, T> &other)
+  RAJA_HOST_DEVICE
+  CudaReduce(const CudaReduce & other)
+#if !defined(__CUDA_ARCH__)
+      : parent{other.parent},
+#else
+      : parent{&other},
+#endif
+        info(other.info),
+        val(other.val)
   {
-    *this = other;
-#if defined(__CUDA_ARCH__)
-    m_is_copy_device = true;
-    m_finish_reduction = !other.m_is_copy_device;
-    extern __shared__ unsigned char sd_block[];
-    T *sd = reinterpret_cast<T *>(&sd_block[m_smem_offset]);
-
-    int threadId = threadIdx.x + blockDim.x * threadIdx.y
-                   + (blockDim.x * blockDim.y) * threadIdx.z;
-
-    // initialize shared memory
-    T val = static_cast<T>(0);
-    for (int i = BLOCK_SIZE / 2; i > 0; i /= 2) {
-      // this descends all the way to 1
-      if (threadId < i) {
-        sd[threadId + i] = val;
+#if !defined(__CUDA_ARCH__)
+    if (parent) {
+      if (val.setupForDevice(info)) {
+        parent = nullptr;
       }
     }
-    if (threadId < 1) {
-      sd[threadId] = val;
-    }
-
-    __syncthreads();
-#else
-    m_is_copy_host = true;
-    m_smem_offset = getCudaSharedmemOffset(m_myID, BLOCK_SIZE, sizeof(T));
 #endif
   }
 
-  /*!
-   * \brief Finish reduction on device and free memory on host.
-   *
-   * Destruction on host releases the device memory chunk for
-   * reduction id and id itself for others to use.
-   * Destruction on device completes the reduction.
-   *
-   * Note: destructor executes on both host and device.
-   */
-  __host__ __device__ ~ReduceSum<cuda_reduce<BLOCK_SIZE, Async>, T>()
+  //! apply reduction on device upon destruction
+  RAJA_HOST_DEVICE
+  ~CudaReduce()
   {
-#if defined(__CUDA_ARCH__)
-    if (m_finish_reduction) {
-      extern __shared__ unsigned char sd_block[];
-      T *sd = reinterpret_cast<T *>(&sd_block[m_smem_offset]);
-
-      int blockId = blockIdx.x + blockIdx.y * gridDim.x
-                    + gridDim.x * gridDim.y * blockIdx.z;
+#if !defined(__CUDA_ARCH__)
+    if (parent == this) {
+      val.cleanup(info);
+    } else if (parent) {
+#if defined(RAJA_ENABLE_OPENMP)
+#pragma omp critical (CudaReduce)
```

Comment:
The downside is with any CudaReduce object, it will create a barrier across all of them (even if they are from completely different reducer object "trees". OMP Locks can be created for each "reducer tree" and allow multiple reducers operating at the same time.